warehouse/oidc/forms.py

```diff
@@ -0,0 +1,121 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+import requests
+import sentry_sdk
+import wtforms
+
+from warehouse import forms
+from warehouse.i18n import localize as _
+
+_VALID_GITHUB_REPO = re.compile(r"^[a-zA-Z0-9-_.]+$")
+
+
+class GitHubProviderForm(forms.Form):
+    __params__ = ["owner", "repository", "workflow_name"]
+
+    owner = wtforms.StringField(
+        validators=[
+            wtforms.validators.DataRequired(
+                message=_("Specify GitHub owner (username or organization)")
+            ),
+        ]
+    )
+
+    repository = wtforms.StringField(
+        validators=[
+            wtforms.validators.DataRequired(message=_("Specify repository slug")),
+            wtforms.validators.Regexp(
+                _VALID_GITHUB_REPO, message=_("Invalid repository name")
+            ),
+        ]
+    )
+
+    workflow_name = wtforms.StringField(
+        validators=[wtforms.validators.DataRequired(message=_("Specify workflow name"))]
+    )
+
+    def __init__(self, *args, api_token, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._api_token = api_token
+
+    def _headers_auth(self):
+        if not self._api_token:
+            return {}
+        return {"Authorization": f"token {self._api_token}"}
+
+    def validate_owner(self, field):
+        owner = field.data
+
+        # To actually validate the owner, we ask GitHub's API about them.
+        # We can't do this for the repository, since it might be private.
+        response = requests.get(
```

Comment:
d0b37d25 refactors this to use exception-driven handling; let me know if that's what you had in mind!

It doesn't 500 the page since I figure it's form-internal logic and might be transient, and so (IMO) we should just spit back a vaguely descriptive error and tell the user to try again momentarily. But let me know if I've underthought this 🙂

Reply:
I do agree that it *should not* 500 the whole page, and it would have with the previous implementation, butt the way it's done now is that it will only 500 in case of a network error that is not a timeout, which should probably be enough.

Also, if we have a 403 for any other reason, we might get a confusing message, but I'm not sure I see a scenario where we have a 403, though I don't know GitHub's permission model that well.

I think that should be 👍 for me for a start, and we'll always be able to make it better on the next iteration 😀